Python 3 Support

nmurtagh opened the pull request:
Addresses #13

nmurtagh committed:
Python 3 Support

nmurtagh commented on tests.py:
```diff
@@ -365,18 +366,21 @@ def test_b64encode_url_utf8(self):
+        self.assertEqual(base64_encoded[-1], '=')
 
         # actual test
-        self.assertNotEqual(jose.b64encode_url(istr), '=')
```
This assertion looks broken

Newky commented on jose.py:
```diff
@@ -305,27 +315,32 @@ def b64decode_url(istr):
-        # this will fail if istr is already encoded
-        pass
-    return istr
+if six.PY3:
```
we might want to refactor this style: http://astrofrog.github.io/blog/2016/01/12/stop-writing-python-4-incompatible-code/

nmurtagh replied:
Good point. I'm hoping I can come up with a way to avoid these sections altogether, but for now I will switch to that style. Thanks.

Newky commented on setup.py:
```diff
@@ -42,7 +48,8 @@ def finalize_package_data(self):
-    setup(name=pkg_name,
+    setup(
+        name=pkg_name,
         version='1.0.0',
```
version bump?

Newky commented on setup.py:
```diff
@@ -58,9 +65,10 @@ def finalize_package_data(self):
-        entry_points = {
+            'Topic :: Software Development :: Libraries'
+        ],
+        # cmdclass={'bdist_rpm': bdist_rpm},
```
why did this get commented out?

nmurtagh replied:
Was debugging something, it's back now :)

Newky commented:
Looks good to me

nmurtagh committed:
Changes after review

yuriikonovaliuk commented on jose.py:
```diff
@@ -139,14 +137,15 @@ def encrypt(claims, jwk, adata='', add_header=None, alg='RSA-OAEP',
-        ('enc', enc), ('alg', alg)])
+    header = dict(
+        list((add_header or {}).items()) + [('enc', enc), ('alg', alg)]
+    )
```
You didn't add this but it looks a bit ugly.
It would look much better this way: `header = dict(add_header or {}, enc=enc, alg=alg)`

yuriikonovaliuk commented on jose.py:
```diff
@@ -257,11 +266,12 @@ def sign(claims, jwk, add_header=None, alg='HS256'):
     (hash_fn, _), mod = JWA[alg]
 
-    header = dict((add_header or {}).items() + [('alg', alg)])
+    header = dict(list((add_header or {}).items()) + [('alg', alg)])
```
Same here.

yuriikonovaliuk commented on tests.py:
```diff
@@ -24,12 +25,14 @@
-        ('enc', enc), ('alg', alg)])
+    header = dict(
+        list((add_header or {}).items()) + [('enc', enc), ('alg', alg)]
+    )
```
The same `dict` creation magic.

yuriikonovaliuk commented:
Please add a note to `CHANGES` file.

nmurtagh commented:
@Newky @yuriikonovaliuk Thanks for the reviews guys - I added a commit with some cleanups, please take a look. I removed `encode_safe` as it didn't seem to do anything useful...

nmurtagh committed:
flake8

hmpf commented on setup.py:
```diff
@@ -58,9 +66,10 @@ def finalize_package_data(self):
             'Operating System :: OS Independent',
             'Programming Language :: Python :: 2 :: Only',
```
This line you no longer need...

yuriikonovaliuk replied:
I guess you can just change it to `Programming Language :: Python :: 2.7` and add some more with Python 3.* versions

hmpf commented on jose.py:
```diff
@@ -1,49 +1,31 @@
-    logger.warn('cjson not found, falling back to stdlib json')
-    from json import loads as json_decode, dumps as json_encode
-
+import six
```
You only seem to be using six.b, so normally I would just pull out six.b and have it as a function locally.

hmpf commented:
Splendid! Django is dropping support for 2.7 next year and jose was one of the hold outs.

jordanambra commented:
Will this get its merge conflicts resolved and an update pushed to pypi?

bmamouri commented:
Why the maintainer of this project hate Python 3 so much?

This should have already been merged into master.

PegasusWang commented:
Plz,  push to pypi?

Jamim commented:
Hi @nmurtagh,
Could you please resolve the conflicts?
Thank you!

IamTugy commented:
is that repository deprecated? it doesnt work with python 3.
get this error right at the init:
`  File "/Users/*****/*****/*****.py", line 8, in <module>
    from jose import jwt  # type: ignore
    ^^^^^^^^^^^^^^^^^^^^
  File "/Users/*****/Library/Caches/pypoetry/virtualenvs/*****-py3.11/lib/python3.11/site-packages/jose.py", line 546
    print decrypt(deserialize_compact(jwt), {'k':key},
    ^^^^^^^^^^^^^^^^^^^^^^^^^^
SyntaxError: Missing parentheses in call to 'print'. Did you mean print(...)?`
